src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/util/SSLConnectionTestUtil.java

```diff
@@ -0,0 +1,183 @@
+     * - Send Client Hello to check if the server replies with Server Hello which indicates that Server understands SSL
+     * - Send ES Ping to check if the server replies to the ES Ping message
+     *
+     * @return SSLConnectionTestResult i.e. ES_PING_FAILED or SSL_NOT_AVAILABLE or SSL_AVAILABLE
+     */
+    public SSLConnectionTestResult testConnection() {
+        if (sendDualSSLClientHello()) {
+            return SSLConnectionTestResult.SSL_AVAILABLE;
+        }
+
+        if (sendESPing()) {
+            return SSLConnectionTestResult.SSL_NOT_AVAILABLE;
+        }
+
+        return SSLConnectionTestResult.ES_PING_FAILED;
+    }
+
+    private boolean sendDualSSLClientHello() {
+        boolean dualSslSupported = false;
+        Socket socket = null;
+        try {
+            OutputStreamWriter outputStreamWriter;
+            InputStreamReader inputStreamReader;
+            if(testSocket != null) {
+                socket = testSocket;
+                outputStreamWriter = testOutputStreamWriter;
+                inputStreamReader = testInputStreamReader;
+            } else {
+                socket = new Socket(host, port);
+                outputStreamWriter = new OutputStreamWriter(socket.getOutputStream(), StandardCharsets.UTF_8);
+                inputStreamReader = new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8);
+            }
+
+            outputStreamWriter.write("DUALCM");
+            outputStreamWriter.flush();
+            logger.debug("Sent DualSSL Client Hello msg to {}", host);
+
+            StringBuilder sb = new StringBuilder();
+            int currentChar;
+            while ((currentChar = inputStreamReader.read()) != -1) {
+                sb.append((char) currentChar);
+            }
+
+            if (sb.toString().equals("DUALSM")) {
+                logger.debug("Received DualSSL Server Hello msg from {}", host);
+                dualSslSupported = true;
+            }
+        } catch (IOException e) {
+            logger.debug("DualSSL client check failed for {}, exception {}", host, e.getMessage());
+        } finally {
+            logger.debug("Closing DualSSL check client socket for {}", host);
+            if (socket != null && socket.isConnected()) {
+                try {
+                    socket.close();
+                } catch (IOException e) {
+                    logger.error("Exception occurred while closing DualSSL check client socket for {}. Exception: {}", host, e.getMessage());
+                }
+            }
+        }
+        logger.debug("dualSslClient check with server {}, server supports ssl = {}", host, dualSslSupported);
+        return dualSslSupported;
+    }
+
+    private boolean sendESPing() {
+        boolean pingSucceeded = false;
+        Socket socket = null;
+        try {
+            if(testSocket != null) {
+                socket = testSocket;
+            } else {
+                socket = new Socket(host, port);
```

Comment:
do we need to set timeout for the same ?

Reply:
Server closes the connection immediately when it does not recognize the message. Would it still be better to have an explicit read/connection timeout set?